google-cloud-storage/lib/google/cloud/storage/file/signer_v4.rb

```diff
@@ -156,25 +152,48 @@ def canonical_query query, algorithm, credential, goog_date, expires,
             query["X-Goog-Credential"] = credential
             query["X-Goog-Date"] = goog_date
             query["X-Goog-Expires"] = expires
-            query["X-Goog-SignedHeaders"] = CGI.escape signed_headers_str
-            query = query.sort_by { |k, _| k.to_s.downcase }.to_h
+            query["X-Goog-SignedHeaders"] = signed_headers_str
+            query = query.map { |k, v| [escape_query_param(k), escape_query_param(v)] }.to_h
+            query = query.sort.to_h
             canonical_query_str = ""
             query.each { |k, v| canonical_query_str += "#{k}=#{v}&" }
             canonical_query_str.chomp "&"
           end
 
+          ##
+          # Only the characters in the regex set [A-Za-z0-9.~_-] must be left un-escaped; all others must be
+          # percent-encoded using %XX UTF-8 style.
+          def escape_query_param str
+            CGI.escape(str.to_s).gsub("%7E", "~")
+          end
+
+          def host_name virtual_hosted_style, bucket_bound_hostname
+            return bucket_bound_hostname if bucket_bound_hostname
+            virtual_hosted_style ? "#{@bucket_name}.storage.googleapis.com" : "storage.googleapis.com"
+          end
+
           ##
           # The URI-encoded (percent encoded) external path to the file.
-          def ext_path
-            path = "/#{@bucket_name}"
+          def ext_path path_style
+            path = ""
+            path += "/#{@bucket_name}" if path_style
             path += "/#{String(@file_name)}" if @file_name && !@file_name.empty?
-            Addressable::URI.escape path
+            CGI.escape(path).gsub "%2F", "/"
```

Comment:
One way to avoid mutable strings:

```ruby
path_elems = []
path_elems << CGI.escape(@bucket_name.to_s) if path_style
path_elems << CGI.escape(@file_name.to_s) if @file_name && !@file_name.empty?
path_elems.join("/")
```

Reply:
Fixed.